README.md
# Expensive Concrete for JD

This is still a work in progress,  if/when I find some better graphics for the tiles.

Plascrete  (170% walking speed)
    refined-concrete
    steel-plate
    plastic-bar
    water

Plascrete, has coloured versions


Tefloncrete  (200% walking speed)
    plascrete
    lubricant


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.1.2
Date: 2023-01-04
  Bugfixes:
    - removed dump of data.raw that was left in the code by mistake from some debugging.
---------------------------------------------------------------------------------------------------
Version: 1.1.1
Date: 2022-12-03
  Changes:
    - Updated locale discription to mention pressing R will rotate plascrete colours
    - Added MAP setting to enable/disable deconstruction of trees/rocks/cliffs above Expensive Concrete
    - Added STARTUP setting to allow admin to change the amount of Plascrete needed to make Coloured Plascrete
    - Added the ingame coloured refined concrete so that it can be made,  along with the ability to press R to rotate though the colours
---------------------------------------------------------------------------------------------------
Version: 1.1.0
Date: 2020-11-23
  Changes:
    - Added support for 1.1.0
---------------------------------------------------------------------------------------------------
Version: 1.0.2
Date: 2020-10-16
  Changes:
    - Changed how I capture the rotation action of the concrete to be able to cycle through all the colours.
---------------------------------------------------------------------------------------------------
Version: 1.0.0
Date: 2020-10-12
  Changes:
    - Rotating coloured "plascrete" changes its color to one of the next 3 in the sequence below.   Tile are limited to 4x directions.
      -  plascrete-red --> plascrete-green
      -  plascrete-green --> plascrete-blue
      -  plascrete-blue --> plascrete-orange
      -  plascrete-orange --> plascrete-yellow
      -  plascrete-yellow --> plascrete-pink
      -  plascrete-pink --> plascrete-purple
      -  plascrete-purple --> plascrete-brown
      -  plascrete-brown --> plascrete-cyan
      -  plascrete-cyan --> plascrete-acid
      -  plascrete-acid --> plascrete-red
---------------------------------------------------------------------------------------------------
Version: 0.0.1
Date: 2020-03-
  Features:
    - First release


data.lua
local constant = require("constant")
local colors = constant.colors

require("prototypes.item-subgroup")
require("prototypes.tile")
require("prototypes.item")
require("prototypes.recipe")
require("prototypes.technology")

local concrete_tile = data.raw.tile["plascrete"]
local concrete_item = data.raw.item["plascrete"]
local concrete_recipe = data.raw.recipe["plascrete"]
local tech = data.raw.technology["concrete-2"]

for color_name, color in pairs(colors) do
    local new_tile = table.deepcopy(concrete_tile)
    new_tile.name = new_tile.name .. "-" .. color_name
    new_tile.localised_name = { "", { "color." .. color_name }, " ", { "tile-name." .. concrete_tile.name } }
    new_tile.minable.result = new_tile.name
    new_tile.tint = color.chat_color
    new_tile.order = "a-e-" .. color_name
    new_tile.layer = concrete_tile.layer + (color.i * 3)
    new_tile.map_color = color.player_color
    data:extend { new_tile }

    local new_item = table.deepcopy(concrete_item)
    new_item.name = new_item.name .. "-" .. color_name
    new_item.localised_name = { "", { "color." .. color_name }, " ", { "tile-name." .. concrete_tile.name } }
    new_item.place_as_tile.result = new_item.name
    new_item.icons = { { icon = "__ExpensiveConcrete__/graphics/icons/concrete.png", icon_size = 64, icon_mipmaps = 1, tint = color.chat_color } }
    new_item.order = concrete_tile.name .. '-' .. color_name
    data:extend { new_item }

    local new_recipe = table.deepcopy(concrete_recipe)
    new_recipe.name = new_recipe.name .. "-" .. color_name
    new_recipe.localised_name = { "", { "color." .. color_name }, " ", { "tile-name." .. concrete_tile.name } }
    new_recipe.ingredients = { { "plascrete", settings.startup["ExpensiveConcrete-coloured-plascrete-cost"].value } }
    new_recipe.result_count = 1
    new_recipe.result = new_recipe.name
    new_recipe.order = concrete_tile.name .. '-' .. color_name
    data:extend { new_recipe }

    table.insert(tech.effects, { recipe = new_recipe.name, type = "unlock-recipe" })
end

local refined_concrete_tile = data.raw.tile["refined-concrete"]
local refined_concrete_item = data.raw.item["refined-concrete"]
local refined_concrete_recipe = data.raw.recipe["refined-concrete"]
for color_name, color in pairs(colors) do
    local tile = data.raw.tile[color_name .. "-" .. refined_concrete_tile.name]
    tile.minable = table.deepcopy(refined_concrete_tile.minable)
    tile.minable.result = tile.name

    local new_item = table.deepcopy(refined_concrete_item)
    new_item.name = color_name .. "-" .. new_item.name
    new_item.localised_name = { "", { "color." .. color_name }, " ", { "tile-name." .. refined_concrete_tile.name } }
    new_item.place_as_tile.result = new_item.name
    new_item.icons = { { icon = refined_concrete_item.icon, icon_size = 64, icon_mipmaps = 4, tint = color.chat_color } }
    new_item.order = refined_concrete_tile.name .. '-' .. color_name
    data:extend { new_item }

    local new_recipe = table.deepcopy(refined_concrete_recipe)
    new_recipe.name = color_name .. "-" .. new_recipe.name
    new_recipe.localised_name = { "", { "color." .. color_name }, " ", { "tile-name." .. refined_concrete_tile.name } }
    new_recipe.ingredients = { { "refined-concrete", settings.startup["ExpensiveConcrete-coloured-refined-concrete-cost"].value } }
    new_recipe.result_count = 1
    new_recipe.result = new_recipe.name
    new_recipe.order = refined_concrete_tile.name .. '-' .. color_name
    data:extend { new_recipe }

    table.insert(tech.effects, { recipe = new_recipe.name, type = "unlock-recipe" })
end



data:extend(
    {
        {
            type = "custom-input",
            name = "concrete-rotate",
            key_sequence = "R"
        }
    }
)


control.lua
local constant = require("constant")
local colors = constant.colors

local function starts_with(str, start)
    return str:sub(1, #start) == start
end

local function OnConfigurationChanged(e)
    if e.mod_changes and e.mod_changes["ExpensiveConcrete"] then
        for _, force in pairs(game.forces) do
            local recipes = force.recipes
            local tech = "concrete-2"
            if force.technologies[tech] and force.technologies[tech].researched then
                for _, effect in pairs(force.technologies[tech].effects) do
                    if effect.type and effect.type == "unlock-recipe" then
                        if recipes[effect.recipe] then
                            recipes[effect.recipe].enabled = true
                            recipes[effect.recipe].reload()
                        end
                    end
                end
            end
        end
    end
end

local function order_deconstruction(surface, force, position, typ)
    local tile = surface.find_entities_filtered {
        position = position,
        radius = 1,
        type = typ
    }
    if tile and #tile > 0 then
        for _, entity in pairs(tile) do
            if entity.minable then
                entity.order_deconstruction(force)
            end
        end
    end

end

local function RobotTileBuilt(e)
    if not (settings.global["ExpensiveConcrete-decom_crap"].value) then return end
    if e.tile.name ~= "tefloncrete" and not (starts_with(e.tile.name, "plascrete")) then return end

    local surface = game.get_surface(e.surface_index)

    if surface then
        for _, tile in pairs(e.tiles) do
            order_deconstruction(surface, e.robot.force, tile.position, { "tree", "simple-entity", "cliff" })
        end
    end
end

local function PlayerTileBuilt(e)
    if not (settings.global["ExpensiveConcrete-decom_crap"].value) then return end
    if e.tile.name ~= "tefloncrete" and not (starts_with(e.tile.name, "plascrete")) then return end

    local surface = game.get_surface(e.surface_index)
    local player = game.get_player(e.player_index)

    if surface and player then
        for _, tile in pairs(e.tiles) do
            order_deconstruction(surface, player.force, tile.position, { "tree", "simple-entity", "cliff" })
        end
    end
end

local function FlyingText(text, position, color)
    color = color or { r = 1, g = 1, b = 1 }

    -- local surface = game.surfaces.nauvis
    return game.surfaces.nauvis.create_entity {
        name = "tutorial-flying-text",
        text = text,
        position = { x = position.x, y = position.y - 1 },
        color = color
    }
end

script.on_event(
    "concrete-rotate",
    function(e)
        local player = game.players[e.player_index]
        if player.cursor_stack and player.cursor_stack.valid_for_read then
            local color, name
            if string.match(player.cursor_stack.name, "plascrete%-(%a+)") then
                color = string.match(player.cursor_stack.name, "plascrete%-(%a+)")
                name = "plascrete"
            end
            if string.match(player.cursor_stack.name, "(%a+)%-refined%-concrete") then
                color = string.match(player.cursor_stack.name, "(%a+)%-refined%-concrete")
                name = "refined-concrete"
            end
            if colors[color] then
                FlyingText(colors[color].next, player.position, colors[colors[color].next].chat_color)
                if name == "plascrete" then name = "plascrete-" .. colors[color].next end
                if name == "refined-concrete" then name = colors[color].next .. "-refined-concrete" end
                player.cursor_stack.set_stack {
                    name = name,
                    count = player.cursor_stack.count
                }
            end
        end
    end
)

script.on_event(defines.events.on_player_built_tile, PlayerTileBuilt)
script.on_event(defines.events.on_robot_built_tile, RobotTileBuilt)
script.on_configuration_changed(OnConfigurationChanged)


prototypes/item.lua
data:extend(
    {
        {
            type = "item",
            name = "plascrete",
            icon = "__ExpensiveConcrete__/graphics/icons/concrete.png",
            icon_size = 64,
            icon_mipmaps = 1,
            subgroup = "ExpensiveConcrete",
            order = "00",
            stack_size = 100,
            place_as_tile = {
                result = "plascrete",
                condition_size = 1,
                condition = {"water-tile"}
            }
        },
        {
            type = "item",
            name = "tefloncrete",
            icon = "__ExpensiveConcrete__/graphics/icons/teflon3.png",
            icon_size = 64,
            icon_mipmaps = 1,
            subgroup = "ExpensiveConcrete",
            order = "00",
            stack_size = 100,
            place_as_tile = {
                result = "tefloncrete",
                condition_size = 1,
                condition = {"water-tile"}
            }
        }
    }
)


prototypes/tile.lua
local refined_concrete_entity = data.raw["tile"]["refined-concrete"]

data:extend(
    {
        {
            type = "tile",
            name = "plascrete",
            needs_correction = false,
            transition_merges_with_tile = "refined-concrete",
            minable = { hardness = 0.2, mining_time = 0.5, result = "plascrete" },
            mined_sound = refined_concrete_entity.mined_sound,
            collision_mask = { "ground-tile" },
            walking_speed_modifier = refined_concrete_entity.walking_speed_modifier + 0.2,
            layer = refined_concrete_entity.layer,
            decorative_removal_probability = 1,
            variants = {
                main = refined_concrete_entity.variants.main,
                inner_corner_mask = refined_concrete_entity.variants.inner_corner_mask,
                outer_corner_mask = refined_concrete_entity.variants.outer_corner_mask,
                side_mask = refined_concrete_entity.variants.side_mask,
                u_transition_mask = refined_concrete_entity.variants.u_transition_mask,
                o_transition_mask = refined_concrete_entity.variants.o_transition_mask,
                material_background = {
                    picture = "__ExpensiveConcrete__/graphics/terrain/concrete.png",
                    count = 8,
                    hr_version = {
                        picture = "__ExpensiveConcrete__/graphics/terrain/hr-concrete.png",
                        count = 8,
                        scale = 0.5
                    }
                }
            },
            walking_sound = refined_concrete_entity.walking_sound,
            map_color = { r = 60, g = 60, b = 60 },
            pollution_absorption_per_second = 0,
            vehicle_friction_modifier = refined_concrete_entity.vehicle_friction_modifier + 0.2
        },
        {
            type = "tile",
            name = "tefloncrete",
            needs_correction = false,
            -- transition_merges_with_tile = "refined-concrete",
            minable = { hardness = 0.2, mining_time = 0.5, result = "tefloncrete" },
            mined_sound = refined_concrete_entity.mined_sound,
            collision_mask = { "ground-tile" },
            walking_speed_modifier = 2,
            layer = refined_concrete_entity.layer * 2,
            decorative_removal_probability = 1,
            variants = {
                main = refined_concrete_entity.variants.main,
                inner_corner_mask = refined_concrete_entity.variants.inner_corner_mask,
                outer_corner_mask = refined_concrete_entity.variants.outer_corner_mask,
                side_mask = refined_concrete_entity.variants.side_mask,
                u_transition_mask = refined_concrete_entity.variants.u_transition_mask,
                o_transition_mask = refined_concrete_entity.variants.o_transition_mask,
                material_background = {
                    picture = "__ExpensiveConcrete__/graphics/terrain/teflon512.png",
                    count = 1,
                    hr_version = {
                        picture = "__ExpensiveConcrete__/graphics/terrain/teflon512.png",
                        count = 1,
                        scale = 0.5
                    }
                }
            },
            walking_sound = refined_concrete_entity.walking_sound,
            map_color = { r = 0.1, g = 0.1, b = 0.1, a = 0.1 },
            pollution_absorption_per_second = 0,
            vehicle_friction_modifier = 2
        }
    }
)


constant.lua
local constant = {}

constant.colors = {
    ["red"] = { i = 1, next = "green", player_color = { r = 0.815, g = 0.024, b = 0.0, a = 0.5 }, chat_color = { r = 1.000, g = 0.266, b = 0.241 } },
    ["green"] = { i = 2, next = "blue", player_color = { r = 0.093, g = 0.768, b = 0.172, a = 0.5 }, chat_color = { r = 0.173, g = 0.824, b = 0.250 } },
    ["blue"] = { i = 3, next = "orange", player_color = { r = 0.155, g = 0.540, b = 0.898, a = 0.5 }, chat_color = { r = 0.343, g = 0.683, b = 1.000 } },
    ["orange"] = { i = 4, next = "yellow", player_color = { r = 0.869, g = 0.5, b = 0.130, a = 0.5 }, chat_color = { r = 1.000, g = 0.630, b = 0.259 } },
    ["yellow"] = { i = 5, next = "pink", player_color = { r = 0.835, g = 0.666, b = 0.077, a = 0.5 }, chat_color = { r = 1.000, g = 0.828, b = 0.231 } },
    ["pink"] = { i = 6, next = "purple", player_color = { r = 0.929, g = 0.386, b = 0.514, a = 0.5 }, chat_color = { r = 1.000, g = 0.720, b = 0.833 } },
    ["purple"] = { i = 7, next = "brown", player_color = { r = 0.485, g = 0.111, b = 0.659, a = 0.5 }, chat_color = { r = 0.821, g = 0.440, b = 0.998 } },
    ["brown"] = { i = 8, next = "cyan", player_color = { r = 0.300, g = 0.117, b = 0.0, a = 0.5 }, chat_color = { r = 0.757, g = 0.522, b = 0.371 } },
    ["cyan"] = { i = 9, next = "acid", player_color = { r = 0.275, g = 0.755, b = 0.712, a = 0.5 }, chat_color = { r = 0.335, g = 0.918, b = 0.866 } },
    ["acid"] = { i = 10, next = "black", player_color = { r = 0.559, g = 0.761, b = 0.157, a = 0.5 }, chat_color = { r = 0.708, g = 0.996, b = 0.134 } },
    ["black"] = { i = 11, next = "red", player_color = { r = 0.1, g = 0.1, b = 0.1, a = 0.5 }, chat_color = { r = 0.5, g = 0.5, b = 0.5 } },

}

return constant
